c/compare.py

```diff
@@ -0,0 +1,216 @@
+import yaml
+import os
+import shutil
+import subprocess
+import tempfile
+
+# files where preprocessed files are different
+BLACKLIST = ["floats-esbmc-regression/trunc_nondet_2.i", "*pthread*/*"]
+
+# categories to be excluded ... (with reason and debug information)
+CATEGORY_BLACKLIST = {
+  "ConcurrencySafety-Main": "(platform-dependent types)",
+  "Systems-OpenBSD-MemSafety": "(only custom includes, no system headers, complicated build process)",
+  "Systems-SQLite-MemSafety": "(complicated build process, requires patched version of cilly)",
+}
+
+# categories to be excluded, if option "skip-large" is used ... (with reason and debug information)
+LARGE_CATEGORY_BLACKLIST = {
+  "SoftwareSystems-DeviceDriversLinux64-ReachSafety": "(only custom includes, no system headers, checking takes too much time)",
+}
+
+# no original source available, there are only preprocessed files.
+# for LDV: there is a related .cil.c file, but it doesn't necessarily match at all
+# for loops/s3.i: this single file is special
+TASKS_ONLY_PREPROCESSED = ['ddv-machzwd/', 'aws-c-common/', 'ldv-linux-3.0/', 'ldv-regression/', 'loops/s3.i']
+
+CBMC_GIT_PATH = "../cbmc.git/"
+
+
+def fail(*msg):
+  print("CRITICAL ERROR:", *msg)
+  exit(1)
+
+
+def expand(s):
+  return sorted(glob.glob(s))
+
+
+def get_architecture(setname):
+  cfgfile = setname + ".cfg"
+  if not os.path.exists(cfgfile):
+    fail("No .cfg file present for category", setname)
+
+  with open(cfgfile, "r") as fp:
+    for line in fp:
+      if line.startswith("Architecture"):
+        bits = line.split()[1]
+        if bits in ["32", "64"]:
+          return bits
+        fail("Invalid bit width in file", setname + ".cfg")
+  fail("Invalid configuration file", cfgFile)
+
+
+def get_tasks_from_set(setFile):
+  with open(setFile, "r") as fp:
+    for line in fp:
+      if not line.startswith("#"): # ignore comments
+        for task in expand(line.strip()):
+          yield task
+
+
+def build_goto_cc():
+  """ build goto-cc and goto-diff if not available, and then set PATH to find it """
+  if not shutil.which("goto-cc") or not shutil.which("goto-diff"):
+    if not os.path.exists(CBMC_GIT_PATH + "src/goto-cc/goto-cc"):
+      subprocess.check_call(["git", "clone", "--depth=1", "http://github.com/diffblue/cbmc.git", CBMC_GIT_PATH])
+      subprocess.check_call(["make", "-j2", "minisat2-download"], cwd=CBMC_GIT_PATH + "src")
+      subprocess.check_call(["make", "-j2", "CXX=g++-5", "goto-diff.dir", "goto-cc.dir"], cwd=CBMC_GIT_PATH + "src")
+    cwd = os.getcwd()
+    os.environ['PATH'] = ":".join([
+        cwd + "/" + CBMC_GIT_PATH + "src/goto-cc",
+        cwd + "/" + CBMC_GIT_PATH + "src/goto-diff",
+        os.environ['PATH']
+    ])
+
+
+def execute_goto_cc(args, setfile, bits, orig, taskfile):
+  """ convert both preprocessed and non-preprocessed files into goto-cc intermediate language and compare them """
+  origoutfile = tempfile.NamedTemporaryFile(prefix="compare_orig_", suffix=".out")
```

Comment:
`NamedTemporaryFile` can be used with `with`, which would be cleaner than `try finally` and safer, because currently the code does not cleanup `taskoutfile` if `origoutfile.close()` throws an exception.

That's the only recommendation I have now.